File: day11/day11.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int is_good(const char *s)
{
	int rule1 = 0;
	int rule3 = 0;
	char last = 0;
	const char *pos = s;
	for (; *s; s++)
	{
		/* rule2: 'i' 'o' 'l' forbidden */
		if (strchr("iol", *s))
		{
			return 0;
		}

		/* rule1: increasing straight of at least 3 letters */
		if (rule1 < 2)
		{
			rule1 = (*s == last+1) ? (rule1+1) : 0;
		}

		/* rule3: non-overlapping pairs of letters */
		if (*s == last && s - pos < 2)
		{
			rule3++;
		}
		else
		{
			pos = s;
		}
		last = *s;
	}
	return rule1 >= 2 && rule3 >= 2;
}

static int next_password(char *pw, size_t len)
{
	size_t i = len;
	while (i-->0)
	{
		if (pw[i] != 'z')
		{
			pw[i]++;
			if (is_good(pw))
			{
				return 1;
			}
			/* restart */
			i = len;
		}
		else
		{
			pw[i] = 'a';
		}
	}
	return 0;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		perror("Cannot open the file");
		return 1;
	}

	size_t spassword = 0;
	char *password = NULL;
	int r = getline(&password, &spassword, input);
	fclose(input);
	if (r == -1)
	{
		perror("Cannot parse the data");
		return 1;
	}

	/* right trim */
	size_t len = strlen(password);
	while (len > 0 && isspace(password[len-1]))
	{
		password[--len] = 0;
	}

	if (next_password(password, len))
	{
		printf("Part1: %s\n", password);
	}
	if (next_password(password, len))
	{
		printf("Part2: %s\n", password);
	}
	free(password);
	return 0;
}

File: day11/day11.py
#!/usr/bin/env python3

import sys

def is_good(pwd):
    rule1 = 0
    rule3 = 0
    last = 0
    pos = 0
    for i, x in enumerate(pwd):
        # rule2: iol forbidden
        if x in "iol":
            return False

        # rule1: three consecutive increasing letters
        if rule1 < 2:
            if last and ord(x) == ord(last)+1:
                rule1 += 1
            else:
                rule1 = 0

        # rule3: non-overlapping pairs of letters
        if last == x and i - pos < 2:
            rule3 += 1
        else:
            pos = i

        last = x

    return rule1 >= 2 and rule3 >= 2

def next_pwd(pwd):
    pwd = list(pwd)
    i = l = len(pwd)
    while i > 0:
        i -= 1
        if pwd[i] != "z":
            pwd[i] = chr(ord(pwd[i])+1)
            if is_good(pwd):
                break

            i = l
        else:
            pwd[i] = "a"

    return "".join(pwd)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except OSError as e:
        print(e, file=sys.stderr)
        sys.exit(1)

    pwd = next_pwd(txt)
    print("Part1:", pwd)

    pwd = next_pwd(pwd)
    print("Part2:", pwd)
